### PR TITLE
Self-Healing Overlap Estimator

### PR DESCRIPTION
**Pull request by jtkrogel**
## Proposed changes

Add implementation of self-healing overlap estimator for legacy and batched drivers.  

As requested, a 4 mpi 4 omp-thread deterministic test is included.  The check_stats.py testing script was updated to handle deterministic tests for stat.h5 quantities based on an absolute tolerance check in addition to the statistical check capability already present. 

## What type(s) of changes does this code introduce?

- New feature
- Testing changes (e.g. new unit/integration/performance tests)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

Laptop

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed

### CODE REVIEW

File: src/QMCWaveFunctions/WaveFunctionComponent.h [hunk 1]
```diff
@@ -483,6 +486,12 @@ class WaveFunctionComponent : public QMCTraits
   */
   virtual void evaluateDerivativesWF(ParticleSet& P, const opt_variables_type& optvars, Vector<ValueType>& dlogpsi);
 
+  /** Compute the derivatives of the log of the wavefunction component with respect to the optimizable parameters.
+   *  Unlike evaluateDerivativesWF, knowledge of mappings from component parameters to the full set of wavefunction parameters is not required.
+   */
+  virtual void evaluateDerivativesWF_local(Vector<ValueType>& dlogpsi);
```
Comment (ye-luo) [inline src/QMCWaveFunctions/WaveFunctionComponent.h:492]: Please remove from WFC and keep it only to MSD.

File: src/QMCWaveFunctions/WaveFunctionComponent.h [hunk 2]
```diff
@@ -532,6 +541,12 @@ class WaveFunctionComponent : public QMCTraits
                                    std::vector<ValueType>& ratios,
                                    Matrix<ValueType>& dratios);
 
+  
+  virtual void detRatios(Vector<ValueType>& ratios)
```
Comment (ye-luo) [inline src/QMCWaveFunctions/WaveFunctionComponent.h:545]: Please remove from WFC and keep it only to MSD.

Comment (ye-luo) [reply src/QMCWaveFunctions/WaveFunctionComponent.h:545]: What does this function calculate?

File: tests/molecules/LiH_ae_MSD/det_vmc_sh_batched.in.xml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    <pairpot name="IonIon" type="coulomb" source="ion0" target="ion0"/>
+    <pairpot name="IonElec" type="coulomb" source="ion0" target="e"/>
+  </hamiltonian>
+   <qmc method="vmc" move="pbyp">
+      <parameter name="blocks"              >    3            </parameter>
+      <parameter name="steps"               >    2            </parameter>
+      <parameter name="subSteps"            >    2            </parameter>
+      <parameter name="timestep"            >    0.4          </parameter>
+      <parameter name="warmupSteps"         >    2            </parameter>
+      <estimator type="selfhealingoverlap" name="sh_overlap" />
```
Comment (ye-luo) [inline tests/molecules/LiH_ae_MSD/det_vmc_sh_batched.in.xml:20]: @PDoakORNL is bare estimator still allowed?

File: src/QMCWaveFunctions/WaveFunctionComponent.h (hunk 1 above)
Comment (jtkrogel) [reply src/QMCWaveFunctions/WaveFunctionComponent.h]: This function is unused, removed now.

File: src/QMCWaveFunctions/WaveFunctionComponent.h (hunk 2 above)
Comment (jtkrogel) [reply src/QMCWaveFunctions/WaveFunctionComponent.h]: Removed from WFC, functionality only at MSD level now.

Inline documentation added.

**Comment by jtkrogel** [comment]
@ye-luo I've addressed all of your comments/requests.

**Comment by jtkrogel** [comment]
Not expecting to add more commits.  Ready for merge.

**Review by ye-luo — APPROVED** [review]
Overall OK. I will do some refactoring later to avoid the redundant code added by detRatios.

**Comment by ye-luo** [comment]
Test this please

File: tests/scripts/check_stats.py [hunk 4]
```diff
@@ -1738,6 +1747,56 @@ def load_scalar_file(options,selector):
...
+        ),
+    'momentum'      : obj(
+        default_label = 'nofk',
+        data_paths    = obj(tot='value'),
+        ),
+    'sh_coeff'      : obj(
+        default_label = 'sh_coeff',
+        data_paths    = obj(coeff='value'),
+        ),
+    })
```
Comment (ye-luo) [inline tests/scripts/check_stats.py:1795]: Incorrect mapping in the above caused test failure. Please check against the old table.
```
        quantity_paths = obj({
        qpaths = stat_info[options.quantity].data_paths
            'density'       : obj(tot='value'),
            'spindensity'   : obj(u='u/value',
                                  d='d/value'),
            '1rdm'          : obj(u='number_matrix/u/value',
                                  d='number_matrix/d/value'),
            '1redm'         : obj(u='energy_matrix/u/value',
                                  d='energy_matrix/d/value'),
            'obdm'          : obj(u='number_matrix/u/value',
                                  d='number_matrix/d/value'),
            'energydensity' : obj(W=('spacegrid1/value',0,3),
                                  T=('spacegrid1/value',1,3),
                                  V=('spacegrid1/value',2,3)),
            'momentum'      : obj(tot='value'),
```

**Review by ye-luo — CHANGES_REQUESTED** [review]
Fixing test failure required.

**Comment by jtkrogel** [comment]
Fixed.  Estimator tests are now operable:
```
>ctest -R deterministic-LiH_ae_msdj
Test project /home/j1k/repositories/qmcpack_dev_new/build_real
    Start 611: deterministic-LiH_ae_msdj_legacy-vmc-estimator-sh_coeff-4-4
1/4 Test #611: deterministic-LiH_ae_msdj_legacy-vmc-estimator-sh_coeff-4-4 ..........   Passed    9.29 sec
    Start 612: deterministic-LiH_ae_msdj_legacy-vmc-estimator-sh_coeff-4-4-check
2/4 Test #612: deterministic-LiH_ae_msdj_legacy-vmc-estimator-sh_coeff-4-4-check ....   Passed    0.36 sec
    Start 613: deterministic-LiH_ae_msdj_batched-vmc-estimator-sh_coeff-4-4
3/4 Test #613: deterministic-LiH_ae_msdj_batched-vmc-estimator-sh_coeff-4-4 .........   Passed    7.96 sec
    Start 614: deterministic-LiH_ae_msdj_batched-vmc-estimator-sh_coeff-4-4-check
4/4 Test #614: deterministic-LiH_ae_msdj_batched-vmc-estimator-sh_coeff-4-4-check ...   Passed    0.36 sec

>ctest -R deterministic-diamondC_1x1x1_pp-dmc-estimator
Test project /home/j1k/repositories/qmcpack_dev_new/build_real
      Start 1125: deterministic-diamondC_1x1x1_pp-dmc-estimator-density-1-1
 1/16 Test #1125: deterministic-diamondC_1x1x1_pp-dmc-estimator-density-1-1 .............   Passed    1.21 sec
      Start 1126: deterministic-diamondC_1x1x1_pp-dmc-estimator-density-1-1-check
 2/16 Test #1126: deterministic-diamondC_1x1x1_pp-dmc-estimator-density-1-1-check .......   Passed    0.35 sec
      Start 1127: deterministic-diamondC_1x1x1_pp-dmc-estimator-density-4-4
 3/16 Test #1127: deterministic-diamondC_1x1x1_pp-dmc-estimator-density-4-4 .............   Passed   11.62 sec
      Start 1128: deterministic-diamondC_1x1x1_pp-dmc-estimator-density-4-4-check
 4/16 Test #1128: deterministic-diamondC_1x1x1_pp-dmc-estimator-density-4-4-check .......   Passed    0.36 sec
      Start 1135: deterministic-diamondC_1x1x1_pp-dmc-estimator-spindensity-1-1
 5/16 Test #1135: deterministic-diamondC_1x1x1_pp-dmc-estimator-spindensity-1-1 .........   Passed    1.01 sec
      Start 1136: deterministic-diamondC_1x1x1_pp-dmc-estimator-spindensity-1-1-check
 6/16 Test #1136: deterministic-diamondC_1x1x1_pp-dmc-estimator-spindensity-1-1-check ...   Passed    0.36 sec
      Start 1137: deterministic-diamondC_1x1x1_pp-dmc-estimator-spindensity-4-4
 7/16 Test #1137: deterministic-diamondC_1x1x1_pp-dmc-estimator-spindensity-4-4 .........   Passed   11.43 sec
      Start 1138: deterministic-diamondC_1x1x1_pp-dmc-estimator-spindensity-4-4-check
 8/16 Test #1138: deterministic-diamondC_1x1x1_pp-dmc-estimator-spindensity-4-4-check ...   Passed    0.36 sec
      Start 1145: deterministic-diamondC_1x1x1_pp-dmc-estimator-1rdm-1-1
 9/16 Test #1145: deterministic-diamondC_1x1x1_pp-dmc-estimator-1rdm-1-1 ................   Passed    1.85 sec
      Start 1146: deterministic-diamondC_1x1x1_pp-dmc-estimator-1rdm-1-1-check
10/16 Test #1146: deterministic-diamondC_1x1x1_pp-dmc-estimator-1rdm-1-1-check ..........   Passed    0.36 sec
      Start 1147: deterministic-diamondC_1x1x1_pp-dmc-estimator-1rdm-4-4
11/16 Test #1147: deterministic-diamondC_1x1x1_pp-dmc-estimator-1rdm-4-4 ................   Passed   16.40 sec
      Start 1148: deterministic-diamondC_1x1x1_pp-dmc-estimator-1rdm-4-4-check
12/16 Test #1148: deterministic-diamondC_1x1x1_pp-dmc-estimator-1rdm-4-4-check ..........   Passed    0.37 sec
      Start 1151: deterministic-diamondC_1x1x1_pp-dmc-estimator-momentum-1-1
13/16 Test #1151: deterministic-diamondC_1x1x1_pp-dmc-estimator-momentum-1-1 ............   Passed    1.09 sec
      Start 1152: deterministic-diamondC_1x1x1_pp-dmc-estimator-momentum-1-1-check
14/16 Test #1152: deterministic-diamondC_1x1x1_pp-dmc-estimator-momentum-1-1-check ......   Passed    0.37 sec
      Start 1153: deterministic-diamondC_1x1x1_pp-dmc-estimator-momentum-4-4
15/16 Test #1153: deterministic-diamondC_1x1x1_pp-dmc-estimator-momentum-4-4 ............   Passed   11.62 sec
      Start 1154: deterministic-diamondC_1x1x1_pp-dmc-estimator-momentum-4-4-check
16/16 Test #1154: deterministic-diamondC_1x1x1_pp-dmc-estimator-momentum-4-4-check ......   Passed    0.36 sec
```

File: src/QMCWaveFunctions/Fermion/MultiSlaterDetTableMethod.h [hunk 5]
```diff
@@ -186,6 +187,14 @@ class MultiSlaterDetTableMethod : public WaveFunctionComponent, public Optimizab
 
   void evaluateDerivativesWF(ParticleSet& P, const opt_variables_type& optvars, Vector<ValueType>& dlogpsi) override;
 
+  /** Compute ratios of the individual Slater determinants and the total MSD value.  
+   *  These are obtained via derivatives of the log of the wavefunction component with respect to the optimizable parameters.
+   *  Unlike evaluateDerivativesWF, knowledge of mappings from component parameters to the full set of wavefunction parameters is not required.
+   */
+  void detRatios(Vector<ValueType>& ratios);
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/MultiSlaterDetTableMethod.h:194]: This function doesn't conform naming rules that functions start with a verb in lower case. I would recommend `calcIndividualDetRatios`.
Although calculating individual determinant ratios and calculating parameter derivatives share very similar internal implementation in the case of CI not necessarily the case with CSFs. They are conceptually different from APIs perspective. The difference is larger than just a mapping for example the det 0 is skipped when calculating derivatives. So I would recommend removing
```
 These are obtained via derivatives of the log of the wavefunction component with respect to the optimizable parameters.
   *  Unlike evaluateDerivativesWF, knowledge of mappings from component parameters to the full set of wavefunction parameters is not required.`
```
in the documentation. Please also document the argument `ratios`.

Comment (ye-luo) [reply src/QMCWaveFunctions/Fermion/MultiSlaterDetTableMethod.h:194]: Also please test this function at
```
https://github.com/QMCPACK/qmcpack/blob/2764f973212f911490ff488a44ed011b22678154/src/QMCWaveFunctions/tests/test_multi_slater_determinant.cpp#L112
```
although you'll need to findMSD and cast the type to MSD first.
Please test the first element and a non-first element.

**Review by ye-luo — APPROVED** [review]
I left two requests to change mostly for myself to do some changes in a later PR but feel free to comment.

**Comment by ye-luo** [comment]
Test this please

**Comment by ye-luo** [comment]
Hit one test failure https://github.com/QMCPACK/qmcpack/actions/runs/9117814869/job/25069532490
It is mixed precision. Probably just increase tolerance?

**Comment by jtkrogel** [comment]
Fixed the mixed precision tests.  Should be good to merge now.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Pull request merged**